fiona/ogrext.pyx

```diff
@@ -1270,12 +1270,20 @@ def _remove(path, driver=None):
     """Deletes an OGR data source
     """
     cdef void *cogr_driver
+    cdef void *cogr_ds
     cdef int result
+    cdef char *driver_c
 
     if driver is None:
-        driver = 'ESRI Shapefile'
+        # attempt to identify the driver by opening the dataset
+        cogr_ds = gdal_open_vector(path.encode("utf-8"), 0, None, {})
+        if cogr_ds == NULL:
+            raise IOError("Failed to remove data source {}".format(path))
```

Comment:
How about using `exc_wrap_pointer` here?

Reply:
I tried this but weirdly it didn't work as expected. In the `test_remove_nonexistent` the call to open returns `NULL`, but `CPLGetLastErrorType()` returns 0. This results in a message `SystemError: <built-in function _remove> returned NULL without setting an error`. Not sure what's going on...